internal/controller/streams_retry.go

```diff
@@ -0,0 +1,110 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/exp/constraints"
+)
+
+type backoffOpts struct {
+	Name                 string
+	InitialBackoff       time.Duration
+	MaxBackoff           time.Duration
+	MaxJitterPct         float64
+	SevereErrorThreshold int
+	ExponentialFactor    float64
+	logger               *zap.Logger
+}
+
+type Action func() error
+
+func clamp[T constraints.Ordered](minimum, val, maximum T) T {
+	if val < minimum {
+		return minimum
+	}
+	if val > maximum {
+		return maximum
+	}
+	return val
+}
+
+type state struct {
+	backoff             time.Duration
+	timer               *time.Timer
+	consecutiveFailures int
+	opts                backoffOpts
+}
+
+func exponentialBackoff(opts backoffOpts, action Action) error {
+	s := state{
+		backoff:             opts.InitialBackoff,
+		consecutiveFailures: 0,
+		timer:               time.NewTimer(0),
```

Comment:
Add a comment. This is important and subtle.
```suggestion
		timer:               time.NewTimer(0), // Don't wait before the first attempt to execute the action.
```